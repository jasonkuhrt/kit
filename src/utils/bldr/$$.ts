export * from './constructor.js'
export * from './mutable.js'
export * from './state.js'
export * from './types.js'
