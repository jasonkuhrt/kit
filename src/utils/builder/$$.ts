export * from './constructor.js'
export * from './state.js'
export * from './types.js'
