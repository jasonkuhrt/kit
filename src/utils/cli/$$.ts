export * from './cli.js'
export * as FlagName from './flag-name/$$.js'
