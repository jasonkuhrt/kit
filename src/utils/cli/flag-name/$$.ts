export * as Analyzer from './analyzer.js'
export * as Checks from './checks.js'
export { FlagName } from './name.js'

// Type-level exports
export type { Analyze, Errors, SomeLimits } from './analyzer.types.js'
