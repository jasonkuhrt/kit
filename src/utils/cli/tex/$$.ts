/**
 * Terminal Text Layout Engine
 *
 * Build rich CLI text layouts with a chainable API. Create bordered blocks, tables,
 * and lists with automatic width calculation, padding, and visual alignment.
 *
 * @category CLI/Text Rendering
 *
 * @example
 * ```typescript
 * import { Tex } from '@wollybeard/kit'
 *
 * // Simple text block
 * const output = Tex.Tex()
 *   .text('Hello World')
 *   .render()
 *
 * // Bordered block with padding
 * Tex.Tex()
 *   .block({
 *     border: { top: '-', bottom: '-', left: '|', right: '|' },
 *     padding: { left: 2, right: 2 }
 *   }, 'Important Message')
 *   .render()
 *
 * // Table with headers
 * Tex.Tex()
 *   .table(($) => $
 *     .headers(['Name', 'Age', 'City'])
 *     .row('Alice', '30', 'NYC')
 *     .row('Bob', '25', 'LA')
 *   )
 *   .render()
 *
 * // Nested blocks
 * Tex.Tex()
 *   .block(($) => $
 *     .text('Parent Block')
 *     .block({ padding: { left: 2 } }, 'Nested Block')
 *   )
 *   .render()
 * ```
 */

export type { BlockBuilder } from './chain/block.js'
export { block } from './chain/block.js'
export type { Builder } from './chain/helpers.js'
export type { RootBuilder } from './chain/root.js'

/**
 * Create a root text layout builder.
 *
 * @category CLI/Text Rendering
 * @param parameters - Optional block parameters for the root container, including terminalWidth
 * @returns A root builder with chainable methods and `.render()` to generate output
 *
 * @example
 * ```typescript
 * // Start building and render
 * const output = Tex.Tex()
 *   .text('Hello')
 *   .text('World')
 *   .render()
 *
 * // With explicit terminal width (useful for testing or fixed-width output)
 * Tex.Tex({ terminalWidth: 80 })
 *   .text('Content')
 *   .render()
 *
 * // With other root-level configuration
 * Tex.Tex({ terminalWidth: 120, padding: { left: 2 } })
 *   .text('Content')
 *   .render()
 * ```
 */
export { createRootBuilder as Tex } from './chain/root.js'

/**
 * Render a builder to a string without calling `.render()` on the builder itself.
 * Useful for rendering builders created in utility functions.
 *
 * @category CLI/Text Rendering
 * @param builder - Any tex builder (root, block, table, or list)
 * @returns The rendered text output
 *
 * @example
 * ```typescript
 * const builder = Tex.Tex().text('Hello')
 * const output = Tex.render(builder)
 * // Same as: builder.render()
 * ```
 */
export { render } from './chain/root.js'

export type { TableBuilder } from './chain/table.js'
export type { BlockParameters } from './nodes/block.js'

/**
 * Block node class. Can be used to create blocks directly instead of via builder methods.
 *
 * @category CLI/Text Rendering
 *
 * @example
 * ```typescript
 * // Usually you use the builder API:
 * Tex.Tex().block('content')
 *
 * // But you can create blocks directly:
 * const myBlock = new Tex.Block({ border: { top: '-' } }, 'content')
 * Tex.Tex().block(myBlock).render()
 * ```
 */
export { Block } from './nodes/block.js'
