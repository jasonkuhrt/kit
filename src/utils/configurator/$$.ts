export { Configurator } from './configurator.js'
export type * from './configurator.js'

// Re-export commonly used types from Configurator namespace at the top level for easier access
import type { Configurator as C } from './configurator.js'
export type ApplyConfiguratorInputResolver$Func<
  $Configurator extends C,
  $Current extends $Configurator['normalizedIncremental'],
  $Input extends $Configurator['input'],
> = C.ApplyConfiguratorInputResolver$Func<$Configurator, $Current, $Input>
export type BuilderProviderCallback<$C extends C> = C.BuilderProviderCallback<$C>
export type Builder<$C extends C> = C.Builder<$C>
export type Configuration = C.Configuration
export type DataInput<$C extends C = C> = C.DataInput<$C>
export type InferParameters<$C extends C> = C.InferParameters<$C>
