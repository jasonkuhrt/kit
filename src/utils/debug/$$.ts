export * from './debug.js'
export * from './dump.js'
export * from './trace.js'
