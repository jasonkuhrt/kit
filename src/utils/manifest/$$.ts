/**
 * Barrel export for Manifest module
 */
export * from './manifest.js'
