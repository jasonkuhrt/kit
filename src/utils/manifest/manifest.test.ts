import { Effect, Schema } from 'effect'
import { Schema as S } from 'effect'
import { describe, expect, test } from 'vitest'
import { type ManifestMutable, ManifestSchema, ManifestSchemaMutable, parseMoniker } from './$$.js'

describe('ManifestSchema', () => {
  test('validates a minimal package.json', () => {
    const minimal = {
      name: 'test-package',
      version: '1.0.0',
    }

    const result = Effect.runSync(Schema.decodeUnknown(ManifestSchema)(minimal))
    expect(result).toMatchObject(minimal)
  })

  test('validates a full package.json', () => {
    const full = {
      name: '@org/package',
      version: '2.0.0',
      description: 'A test package',
      main: 'index.js',
      type: 'module',
      scripts: {
        test: 'vitest',
        build: 'tsc',
      },
      dependencies: {
        effect: '^3.0.0',
      },
      devDependencies: {
        vitest: '^1.0.0',
      },
      peerDependencies: {
        react: '^18.0.0',
      },
      optionalDependencies: {
        'optional-lib': '^1.0.0',
      },
      bin: {
        'my-cli': './cli.js',
      },
      files: ['dist', 'README.md'],
      exports: {
        '.': './index.js',
        './sub': './sub/index.js',
      },
      imports: {
        '#internal': './src/internal.js',
      },
      engines: {
        node: '>=20',
        pnpm: '>=8',
      },
      repository: {
        type: 'git',
        url: 'https://github.com/org/package.git',
      },
      keywords: ['test', 'package'],
      author: {
        name: 'Test Author',
        email: 'test@example.com',
      },
      license: 'MIT',
      bugs: {
        url: 'https://github.com/org/package/issues',
      },
      homepage: 'https://example.com',
      private: false,
      workspaces: ['packages/*'],
      packageManager: 'pnpm@8.0.0',
      madge: {
        detectiveOptions: {
          ts: {
            skipTypeImports: true,
          },
        },
      },
    }

    const result = Effect.runSync(Schema.decodeUnknown(ManifestSchema)(full))
    expect(result).toMatchObject(full)
  })

  test('rejects invalid type field', () => {
    const invalid = {
      name: 'test',
      version: '1.0.0',
      type: 'invalid',
    }

    expect(() => Effect.runSync(Schema.decodeUnknown(ManifestSchema)(invalid))).toThrow()
  })

  test('allows string bin field', () => {
    const manifest = {
      name: 'test',
      version: '1.0.0',
      bin: './cli.js',
    }

    const result = Effect.runSync(Schema.decodeUnknown(ManifestSchema)(manifest))
    expect(result.bin).toBe('./cli.js')
  })

  test('allows string repository field', () => {
    const manifest = {
      name: 'test',
      version: '1.0.0',
      repository: 'https://github.com/org/repo.git',
    }

    const result = Effect.runSync(Schema.decodeUnknown(ManifestSchema)(manifest))
    expect(result.repository).toBe('https://github.com/org/repo.git')
  })

  test('allows string author field', () => {
    const manifest = {
      name: 'test',
      version: '1.0.0',
      author: 'John Doe <john@example.com>',
    }

    const result = Effect.runSync(Schema.decodeUnknown(ManifestSchema)(manifest))
    expect(result.author).toBe('John Doe <john@example.com>')
  })

  test('supports mutable fields with mutable schema', () => {
    const manifest = S.decodeUnknownSync(ManifestSchemaMutable)({
      name: 'test',
      version: '1.0.0',
      scripts: {
        test: 'vitest',
      },
    }) as ManifestMutable

    // Should allow mutation
    manifest.scripts!['build'] = 'tsc'
    expect(manifest.scripts!['build']).toBe('tsc')

    // Arrays should be mutable too
    manifest.files = ['src']
    manifest.files.push('dist')
    expect(manifest.files).toEqual(['src', 'dist'])
  })
})

describe('parseMoniker', () => {
  test('parses scoped package name', () => {
    expect(parseMoniker('@org/package')).toEqual({
      org: '@org',
      name: 'package',
    })
  })

  test('parses unscoped package name', () => {
    expect(parseMoniker('package')).toEqual({
      name: 'package',
    })
  })

  test('handles package with slashes but no scope', () => {
    expect(parseMoniker('not-scoped/package')).toEqual({
      name: 'not-scoped/package',
    })
  })
})
