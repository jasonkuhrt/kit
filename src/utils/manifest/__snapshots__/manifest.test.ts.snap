// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Manifest > validates a full package.json 1`] = `
Manifest {
  "author": {
    "email": "test@example.com",
    "name": "Test Author",
  },
  "bin": {
    "my-cli": "./cli.js",
  },
  "bugs": {
    "url": "https://github.com/org/package/issues",
  },
  "dependencies": {
    "effect": "^3.0.0",
  },
  "description": "A test package",
  "devDependencies": {
    "vitest": "^1.0.0",
  },
  "engines": {
    "node": ">=20",
    "pnpm": ">=8",
  },
  "exports": {
    ".": "./index.js",
    "./sub": "./sub/index.js",
  },
  "files": [
    "dist",
    "README.md",
  ],
  "homepage": "https://example.com",
  "imports": {
    "#internal": "./src/internal.js",
  },
  "keywords": [
    "test",
    "package",
  ],
  "license": "MIT",
  "madge": {
    "detectiveOptions": {
      "ts": {
        "skipTypeImports": true,
      },
    },
  },
  "main": "index.js",
  "name": "@org/package",
  "optionalDependencies": {
    "optional-lib": "^1.0.0",
  },
  "packageManager": "pnpm@8.0.0",
  "peerDependencies": {
    "react": "^18.0.0",
  },
  "private": false,
  "repository": {
    "type": "git",
    "url": "https://github.com/org/package.git",
  },
  "scripts": {
    "build": "tsc",
    "test": "vitest",
  },
  "type": "module",
  "version": OfficialRelease {
    "_tag": "SemverOfficialRelease",
    "build": undefined,
    "major": 2,
    "minor": 0,
    "patch": 0,
    "version": Version {
      "build": undefined,
      "major": 2,
      "minor": 0,
      "patch": 0,
      "prerelease": undefined,
      "raw": "2.0.0",
    },
  },
  "workspaces": [
    "packages/*",
  ],
}
`;

exports[`Manifest > validates a minimal package.json 1`] = `
Manifest {
  "name": "test-package",
  "version": OfficialRelease {
    "_tag": "SemverOfficialRelease",
    "build": undefined,
    "major": 1,
    "minor": 0,
    "patch": 0,
    "version": Version {
      "build": undefined,
      "major": 1,
      "minor": 0,
      "patch": 0,
      "prerelease": undefined,
      "raw": "1.0.0",
    },
  },
}
`;
